SAML sign-in and sign-up

mateuszmandera opened the pull request:
Proper PR, replacing the WIP https://github.com/zulip/zulip/pull/13205

I believe I made the adjustments requested there.
The setup instructions in prod_settings surely can be improved... Probably we should write a proper readthedocs doc for this in ``authentication-methods.md``?

One word about testing - I believe ``test_social_auth_backends`` in ``AuthBackendTest`` is essentially superseded by the whole variety of tests that ``SocialAuthBase`` does now? On that assumption, I made tests based on the ``SocialAuthBase`` framework only.

Finally, screenshot:
![image](https://user-images.githubusercontent.com/45007152/65808949-9d30a400-e198-11e9-918e-1ec06673553b.png)

If we want to do something with this zulip-themed SAML icon I put on the button, I think I can have appropriate adjustments made. Unless we want to go with another icon, or just no icon for now.

andersk commented:
The `PROVISION_VERSION` in `version.py` should be bumped for new dependencies.

I don’t think the `parse_cryptographic_key_to_single_line` transformation is needed: the [documentation](https://python-social-auth.readthedocs.io/en/latest/backends/saml.html#required-configuration) says “you *can* omit the first and last lines”, which also implies you can include them.

mateuszmandera commented:
Oh, the version bump, thanks. Hmm, tried on production that indeed it works  without the parsing function.

timabbott commented:
Merged the dependencies commit.

timabbott commented on templates/zerver/config_error.html:
```diff
@@ -80,6 +80,12 @@ <h1 class="lead">Configuration error</h1>
 
+                    {% if saml_error %}
+                    <p>
+                        SAML authentication is eitheir not enabled or misconfigured. See setup instructions
```
Typo in "either" here.

mateuszmandera replied:
fixed

timabbott commented on templates/zerver/config_error.html:
```diff
@@ -80,6 +80,12 @@ <h1 class="lead">Configuration error</h1>
+                    {% if saml_error %}
+                    <p>
+                        SAML authentication is eitheir not enabled or misconfigured. See setup instructions
+                        in <code>/etc/zulip/settings.py</code>.
```
I think we want to move the main SAML setup docs to a section on `docs/production/authentication-methods.md` and link there to match our standard approach for new settings with complex instructions.

mateuszmandera replied:
Put a link there:
```

                    {% if saml_error %}
                    <p>
                        SAML authentication is either not enabled or misconfigured. Have a look at
                        our <a href="https://zulip.readthedocs.io/en/latest/production/authentication-methods.html#SAML">setup guide</a>.
                    </p>
                    {% endif %}
```

timabbott commented on zerver/tests/fixtures/saml/idp.crt:
```diff
@@ -0,0 +1,17 @@
+MIIDojCCAoqgAwIBAgIGAW0svbVqMA0GCSqGSIb3DQEBCwUAMIGRMQswCQYDVQQGEwJVUzETMBEG
```
Just to confirm, I assume these are test cert values that aren't useful for any non-test authentication?

mateuszmandera replied:
Yeah, random keys generated on my computer. The IdP public key was the public key from my test okta setup, which I used because the samlresponse test fixture has signatures matching it, but since due to the mocking trick we're able to skip their validation, I swapped it out for another random key I generated.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -474,7 +477,7 @@ class SocialAuthBase(ZulipTestCase):
         self.user_profile = self.example_user('hamlet')
         self.email = self.user_profile.email
-        self.name = 'Hamlet'
+        self.name = self.user_profile.full_name
```
This `self.name` refactor could probably be extracted as a prep commit separate from the main SAML one.

mateuszmandera replied:
extracted to another commit.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -942,6 +945,151 @@ def test_social_auth_complete_when_base_exc_is_raised(self) -> None:
+        params['next'] = next
+        params['multiuse_object_key'] = multiuse_object_key
+        if len(params) > 0:
+            url += "?%s" % (urllib.parse.urlencode(params),)
```
Is this block identical with the normal `social_auth_test`?  If so I think we should extract this block as a function `self.get_social_auth_http_params()` that returns (url, params, headers) to avoid duplicating this somewhat complex block of code.

mateuszmandera replied:
extracted, as ``prepare_login_url_and_headers`` in a separate commit now.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -942,6 +945,151 @@ def test_social_auth_complete_when_base_exc_is_raised(self) -> None:
+
+        return result
+
+    def generate_saml_response(self, email: str, name: str) -> str:
```
It's probably worth writing a comment string explaining this function's role.

mateuszmandera replied:
Done:
```
    def generate_saml_response(self, email: str, name: str) -> str:
        """
        The samlresponse.txt fixture has a pre-generated SAMLResponse,
        with {email}, {first_name}, {last_name} placeholders, that can
        be filled out with the data we want.
        """
```

timabbott commented on zproject/backends.py:
```diff
@@ -1003,6 +1009,31 @@ def get_verified_emails(self, *args: Any, **kwargs: Any) -> List[str]:
+        more reasonable.
+        If it is accessed with RelayState specified, but missing SAMLResponse,
+        OneLogin_Saml2_Error ends up raised.
+        We need to catch and handle those here.
```
I think we should add a bit more context to this docstring explaining for someone who doesn't know SAML what these conditions are.  E.g. we should explain what "RelayState isn't specified in POST" means (like is that something that'd be caused by a bug, and previously provided an inscrutible error message?  That's sorta what this sounds like, but it's worth spelling out things like that explicitly as they'll be very helpful for anyone trying to understand this code in 6 months)

mateuszmandera replied:
Added an additional paragraph there:
```
        Both parameters should be present if the user came to /complete/saml/ through
        the IdP as intended. The errors can happen if someone simply types the endpoint into
        their browsers, or generally tries messing with it in some ways.
```

timabbott commented on zproject/backends.py:
```diff
@@ -1003,6 +1009,31 @@ def get_verified_emails(self, *args: Any, **kwargs: Any) -> List[str]:
+            return None
+        except OneLogin_Saml2_Error as e:
+            # This error provides a proper description of the issue, so we can simply log str(e)
+            logging.info(str(e))
```
The meaning of the errors above will tell us how these should be logged.  If it's likely a bug in some software, we probably want at least `logging.warning`.  If it's something that would routinely happen without a bug (e.g. someone hitting the `Back` button at the wrong time that recovers automatically), then `logging.debug` or `.info` might be most appropriate.

mateuszmandera replied:
As explained in the extra paragraph added to the comment, these are caused by stuff like the user going directly to ``/complete/saml/``, so I think ``.info`` is right here?

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -185,6 +186,89 @@
+#   chmod 640 /etc/zulip/saml/zulip-private-key.key
+
+# SOCIAL_AUTH_SAML_SP_PUBLIC_CERT = ""
+# SOCIAL_AUTH_SAML_SP_PRIVATE_KEY = ""
```
I think we don't want these commented-out settings here at all if the recommendation is to not enter anything here, as it's just going to be confusing.

mateuszmandera replied:
Agreed, using your suggested changes to prod_settings_template now.

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -185,6 +186,89 @@
+# }
+
+# Important note: Only one IdP is currently supported, don't add
+# any extra IdPs in this setting.
```
I think this note doesn't belong here, just because this `settings.py` file won't be auto-updated when we update the version of Zulip.  Should go in ReadTheDocs instead.

mateuszmandera replied:
Agreed, using your suggested changes to prod_settings_template now (squashed in from the commit you added here).

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -185,6 +186,89 @@
+#         # Do not edit the x509cert attribute below, instead put the IdP's certificate in:
+#         # /etc/zulip/saml/idps/idp_name.crt
+#         # where idp_name is the name of this IdP, as configured above.
+#         "x509cert": "",
```
Would it work better to just not have an `x509cert` attribute listed here, and just have a simpler comment?

mateuszmandera replied:
True, that's what we have now.

timabbott commented on zproject/settings.py:
```diff
@@ -52,6 +52,30 @@ def get_config(section: str, key: str, default_value: Optional[Any]=None) -> Opt
+    if is_line_redundant(lines[-1]):
+        lines = lines[:-1]
+
+    return ''.join(lines)
```
If we don't need this, it'd be nice to not have it, as it's a decent amount of complexity.

mateuszmandera replied:
I got rid of this function, since indeed the SAML modules handle it on their own.

timabbott commented:
I'm doing a quick edit pass to draft the setup docs.  Will likely need some input from you on several things.

timabbott commented:
Feel free to squash my commit once you've read it; kept it separate for readability.  Here's the list of things that need attention:

* For this line in the docs:
> SAML authentication does not support organization creation at this time; but you can disable EmailAuthBackend once you have created the organization.

I haven't actually checked if this is true.

* Is it best to configure on the IdP end first?  My theory was likely yes because you need info from that end to do the Zulip side configuration.

* Can you test that the `chmod` sequence I provided actually works if one just copy-pastes it?  It's totally untested.

* What is the workflow with Okta/Google for how they tell you to generate your server's certs?  Do we want to have a tool that does that for you (and then document how to upload), or do you download the certs?  Our docs should provide the right way to do things.

* How do we want to name the cert files?  For `/etc/zulip/saml/zulip-private-key.key`, the filename is somewhat redundant; could be  `/etc/zulip/saml/zulip-private-key.key`

* We should make `config-error/saml` link to the ReadTheDocs page for how to setup.

* I uncommented the default/example configuration to cut a step from the instructions, and because it won't do anything unless one enables the auth backend, right?  We should figure out how we want to handle this in the development environment.  

* Is `SOCIAL_AUTH_SAML_SP_ENTITY_ID` something users will ever need to change?  If not, we should just set it in `zproject/settings.py` and not have it appear at all.  

* Is this statement I made accurate?

> `displayname` will appear in the SAML authentication flow ...

(And is there anything else that appears in that flow from this configuration?)  I'm also not sure if we can set `name` to anything in that object, and/or if we should recommend something like "zulip" appear in those fields.  E.g. should `displayname` by "Acme Zulip", "Acme", or "Zulip" for most orgs?

timabbott commented:
Overall this looks pretty great; I think we should be close to having something ready for beta testing once the above comments are resolved.

mateuszmandera commented:
@timabbott 
> What is the workflow with Okta/Google for how they tell you to generate your server's certs? Do we want to have a tool that does that for you (and then document how to upload), or do you download the certs? Our docs should provide the right way to do things.

Okay, this is actually an interesting matter, that I initially left for later research and then overlooked. So, according to the main docs https://python-social-auth.readthedocs.io/en/latest/backends/saml.html the SP certificate + key are obligatory settings. But quick testing shows that by default, python-social-auth saml does NOT sign SAMLRequests (which is the main usage of those keys - signing and encryption, which isn't done by default either). 

Quick dive into the underlying ``python3-saml`` library shows that these keys are actually unnecessary unless you specifically add ``authnRequestsSigned`` in the extra security settings ``SOCIAL_AUTH_SAML_SECURITY_CONFIG``, or configure your IdP to encrypt the SAMLResponse (then python3-saml will attempt to decrypt them using the private key).
But by default, since no encryption/signing using them is happening, these settings can be set to ``''`` (I tested on my production setup with okta).

Long story short, this additional signing/encryption stuff with these keys is optional and adds a significant amount of complexity to the setup. My suggestion would to change the docs to have the default setup with the SP keys, and later on we can expand them with info how all this optional stuff can be set up. What do you think?

mateuszmandera commented:
> SAML authentication does not support organization creation at this time; but you can disable EmailAuthBackend once you have created the organization.

Haven't tried yet either, but I imagine this has to be the same as with our github/google backends. Do those not support organization creation?

>     * Is it best to configure on the IdP end first?  My theory was likely yes because you need info from that end to do the Zulip side configuration.

Yes, I think so.

>     * Can you test that the `chmod` sequence I provided actually works if one just copy-pastes it?  It's totally untested.

I tested, they seem fine.


>     * Is `SOCIAL_AUTH_SAML_SP_ENTITY_ID` something users will ever need to change?  If not, we should just set it in `zproject/settings.py` and not have it appear at all.

Hmm, I don't feel familiar enough with things to be sure if this is something people may want to customize, but at least there shouldn't be a need for, so I suppose it's reasonable to move it to ``zproject/settings``.

>     * Is this statement I made accurate?
> 
> 
> > `displayname` will appear in the SAML authentication flow ...
> 
> (And is there anything else that appears in that flow from this configuration?) I'm also not sure if we can set `name` to anything in that object, and/or if we should recommend something like "zulip" appear in those fields. E.g. should `displayname` by "Acme Zulip", "Acme", or "Zulip" for most orgs?

" `displayname` **can** appear in the SAML authentication flow" would be more accurate perhaps, because I believe this will be dependent on the IdP. For okta, it doesn't appear anywhere for me.
It's only used to fill out some XML fields in the metadata and SAMLRequest, but doesn't appear to impact anything.

mateuszmandera commented:
Okay, since the public/private server keys actually aren't necessary in the default setup, I made adjustments in the code so they're not obligatory, adjusted the documentation for that, plus added extra instruction about setting up extra signing/encryption with the keys. And added two tests:
```
    def test_social_auth_no_key(self) -> None:
        """
        Since in the case of SAML there isn't a direct equivalent of CLIENT_KEY_SETTING,
        we override this test, to test for the case where the obligatory
        SOCIAL_AUTH_SAML_ENABLED_IDPS isn't configured.
        """
        account_data_dict = self.get_account_data_dict(email=self.email, name=self.name)
        with self.settings(SOCIAL_AUTH_SAML_ENABLED_IDPS=None):
            result = self.social_auth_test(account_data_dict,
                                           subdomain='zulip', next='/user_uploads/image')
            self.assertEqual(result.status_code, 302)
            self.assertEqual(result.url, self.CONFIG_ERROR_URL)

    def test_saml_auth_works_without_private_public_keys(self) -> None:
        with self.settings(SOCIAL_AUTH_SAML_SP_PUBLIC_CERT='', SOCIAL_AUTH_SAML_SP_PRIVATE_KEY=''):
            self.test_social_auth_success()
```

mateuszmandera commented:
@timabbott I think this concludes this round of changes.

mateuszmandera commented:
>  SAML authentication does not support organization creation at this time; but you can disable EmailAuthBackend once you have created the organization.

I tested, by installing a new zulip instance and following our doc instructions for configuring SAML, before creating the first realm, and then went to realm generation link, and everything worked fine. So I pushed a commit deleting that first instruction from the doc.

![image](https://user-images.githubusercontent.com/45007152/65840364-61bce380-e318-11e9-8675-e694db516d04.png)

timabbott commented:
OK, great!  Looking at this in detail again now.

timabbott commented:
This looks great; I squashed that commit and did a few more edits to the docs.  Here's what I think still needs to happen:
* I'd love to go through the flow of setting up an Okta application E2E trying to follow the docs to check for things we can improve there.  
* I'm still not sure if we should tell them to start on the Zulip server or on the IdP site in terms of the optimal flow, which we'll probably figure out from that process.
* I think the SAML logo thing looks worse than having no logo at all; I could imagine using the Zulip logo, but it might be cleaner to just make our code for rendering the buttons support not having a logo via an option.
* Thinking about "Log in with SAML", I kinda want the "SAML" there to be picked from something in the /etc/zulip/settings.py config; ideally a field we add to the `SOCIAL_AUTH_SAML_ENABLED_IDPS` format that can be the name of the IDP.  This would be more future-proof for a world with multiple IdPs supported.  
* I'm not sure how this should work in a Zulip server hosting multiple organizations.  Potentially the ideal model would be that for e.g. zulipchat.com we have the keys in  `SOCIAL_AUTH_SAML_ENABLED_IDPS` be the `realm.string_id` values, and that's how we control the mapping of which SAML config goes with which organization.  Definitely a follow-up project to actually do that, but does that make sense for how we'd connect it?  If that's our planned model, we might want to make the default value of `idp_name` be `default` or something as a way of specifying the default for realms that don't have that IdP specified to set things up for that model (or maybe have a `default=True` key, or something?).  But we'd need to read some `python-social-auth` code and think a bit to determine if this concept is a good direction.  I could also imagine adding a `realms=[]` field in the dictionary as a way of specifying that, which would be more flexible.  It's I suppose mainly a question of whether we can access random fields from that data structure in our code or not in a place that controls which IdPs can be used.

timabbott commented:
Pushed a couple more commits with some initial work towards developer documentation for it.  

I ran into that `localhost:9992` bug; not sure what the right way to fix that is.

timabbott commented:
OK, done working on this for now.  Note that I changed the `zproject/prod_settings_template` block for the signature setting to be uncommented and one just needs to change `False` to `True` rather than uncommenting something.  Let me know if you think that was the wrong thing to do.

It looks like we just need to figure out that `localhost:9992` issue and then we can do manual testing in the development environment following the instructions I wrote, which would be really nice.

timabbott commented:
@mateuszmandera it'd be great if you can do the next round of edits on this.  I'm OK with shipping with the `localhost:9992` issue in development if we can't find a good solution to it.  That said, I did a bit of investigation, and here's what I found.

The bug is at least in part that `self.strategy.request_port()` disagrees with `self.strategy.request_host` in this bit of upstream code (`grep -R ` in `/srv/zulip-py3-venv` to find).  

```
    def _create_saml_auth(self, idp):                                                                                           
        """Get an instance of OneLogin_Saml2_Auth"""                                                                            
        config = self.generate_saml_config(idp)                                                                                 
        request_info = {                                                                                                        
            'https': 'on' if self.strategy.request_is_secure() else 'off',                                                      
            'http_host': self.strategy.request_host(),                                                                          
            'script_name': self.strategy.request_path(),                                                                        
            'server_port': self.strategy.request_port(),                                                                        
            'get_data': self.strategy.request_get(),                                                                            
            'post_data': self.strategy.request_post(),                                                                          
        }                                                                                                                       
        return OneLogin_Saml2_Auth(request_info, config)                                                                        
```

I at first thought it possible that our `run-dev.py` proxy was doing something wrong here, and tried something like this:

```
diff --git a/tools/run-dev.py b/tools/run-dev.py
index 3ac9d44600..eccbf9f73a 100755
--- a/tools/run-dev.py
+++ b/tools/run-dev.py
@@ -319,6 +319,8 @@ class CombineHandler(BaseWebsocketHandler):
         # type: () -> None
         if 'X-REAL-IP' not in self.request.headers:
             self.request.headers['X-REAL-IP'] = self.request.remote_ip
+        if 'X-FORWARDED_PORT' not in self.request.headers:
+            self.request.headers['X-FORWARDED-PORT'] = proxy_port
         if self.request.headers.get("Upgrade", "").lower() == 'websocket':
             return super().prepare()
         url = transform_url(
diff --git a/zproject/dev_settings.py b/zproject/dev_settings.py
index e20a47ef42..9ff52cdec7 100644
--- a/zproject/dev_settings.py
+++ b/zproject/dev_settings.py
@@ -151,3 +151,6 @@ BILLING_ENABLED = True
 
 # Test Custom TOS template rendering
 TERMS_OF_SERVICE = 'corporate/terms.md'
+
+# run-dev.py proxies the port.
+USE_X_FORWARDED_PORT = True
```

After remembering one needs to restart `run-dev.py` for changes to its own code, that fixed our first issue.  But afterwards, we get this error:
```
2019-10-07 04:17:43.072 INFO [] Authentication failed: SAML login failed: ['invalid_response'] (https://zulip.example.com is not a valid audience for this Response)
```

Reading and print-debugging `/srv/zulip-py3-venv/lib/python3.6/site-packages/onelogin/saml2/response.py`, it became clear the problem was with how the entity was set in the development environment; I added a line to `dev_settings.py` to address this.

timabbott commented:
I pushed back to the PR with the fixes for the development environment localhost:9992 issue I described above.  The status quo is that SAML auth still doesn't work: We get an `invalid_realm` error, apparently because `self.strategy.session_get("subdomain")` is returning `None` in this code path.  It's not clear to me what the cause is, but we can now at least manually test most of the flow in the development environment.

mateuszmandera commented:
Oh, this is a bug actually.
```
 "GET /login/saml/?subdomain=zulip&is_signup=0&multiuse_object_key=&idp=idp_name
```
starts the auth flow, we get redirected to okta, authenticate there, get redirected to /complete/saml/ - all these GET params (in particular, subdomain) don't get passed back on the return from okta. In hindsight, I totally should realized that earlier - the SAML way of transporting such extra information is in ``RelayState`` and python-social-auth uses RelayState to transport ``idp_name`` only.

The solution is to do a bit more overriding of the built-in methods of SAMLBackend to encode the GET parameters into the RelayState when redirecting to okta, and on the comeback from okta to move them into session params and pass a stripped RelayState (with only the ``idp_name``) to the inside of python-social-auth flow. I'll work on that now.

mateuszmandera commented:
@timabbott Fixed. Things work in development now too. Can you review? I'll work on the previously requested, more minor edits next.

In particular, the params we pass through RelayState, especially the mobile otp - since it'll pass through the IdP, is it fine to send that unencrypted?

timabbott commented:
Interesting question.  The `mobile_flow_otp` exists to protect against attacks by a malicious app that registers a competing `zulip://` URL handler on the user's device (mobile platform's don't make it possible to guarantee the Zulip app exclusively handles requests to `zulip://` URLs) by encrypting the response.  Is it correct to assume that any reasonable SAML provider would be using HTTPS for its communication?   In that case we're only potentially sharing `mobile_flow_otp` with the SAML provider if it's HTTPS isn't compromised.  So for it to be relevant the attacker would need to both control the app handling `zulip://` URLs on the device and the SAML provider (at least at the level of reading RelayState).  Potentially of more concern is passing through the `next` value.  

Probably not much of issue given one needs access to the SAML provider.   We could probably avoid it, though, by sending an opaque random ID to the SAML provider in RelayState and using that as a key in a short-lived key/value setup (e.g. in redis) to look up the parameters we're encoding in RelayState here.

(Also, if I'm not mistaken, this state will be encrypted if SP certs are setup; that's just not available with every provider?  I guess another reason to recommend setting that up if possible.)

mateuszmandera commented:
>  this state will be encrypted if SP certs are setup; that's just not available with every provider?

Actually no, the encryption only applies to the SAMLResponse parameter, RelayState is just passed as it is.

>  Is it correct to assume that any reasonable SAML provider would be using HTTPS for its communication?

I believe so, though in theory SAML doesn't strictly require https.

But yeah, using redis to store the parameters and just sending a token through RelayState allows skipping over all these question marks and should be easy to do, so probably no excuse not to do it that way.

timabbott commented:
OK, I did a few cleanups (mostly to comments) in the last commit.  There's a few things I'd like to explore as new term follow-ups (noted below), but I think it makes sense to merge at this point.  I've done so after squashing the last several commits.  TODOs in https://github.com/zulip/zulip/issues/13275.

Huge thanks for making this happen @mateuszmandera!
